Licensing rework

SoftAdviceJR opened the pull request:
The static methods weren't cutting it for me, because I somehow needed to change the way the types were filtered. I did something that resembles the decorator pattern in that a LicensedDispatcher both inherits from the abstract Dispatcher and wraps an implemented Dispatcher.

I also merged all the different projects into one because it was getting kinda unwieldy and I was starting to get the the point where I was needing bidirectional references between projects.

SoftAdviceJR committed:
Licensing rework

The static methods weren't cutting it for me, because I somehow needed to change the way the types were filtered. I did something that resembles the decorator pattern in that a LicensedDispatcher both inherits from the abstract Dispatcher and wraps an implemented Dispatcher.

I also merged all the different projects into one because it was getting kinda unwieldy and I was starting to get the the point where I was needing bidirectional references between projects.

Co-Authored-By: Frederik Madsen <FredTheDane@users.noreply.github.com>

SoftAdviceJR commented:
There's a lot of stuff I'm not sure about. 
I added some readme's you can look over for usage. 
I'll try to add some unit tests tomorrow.

SoftAdviceJR commented:
So apparently I suck at github and now I've got a bunch of conflicts to solve. woops

SoftAdviceJR commented:
As I'm merging our ideas and learning more about the command pattern, this method I've cooked up where commands are passed in the constructor doesn't seem like such a good idea 😒. I wanted the Dispatch method to be without parameters so it could feasibly be the same on every dispatcher, but I'm starting to doubt that was the way to go.

For now I'll just get everything working again, but we might want to rethink the method signatures.

SoftAdviceJR committed:
Licensing rework

I tried merging our ideas.

- I moved when the callingassembly is added to the assemblies because it was causing troubles. It's now added in the constructor of the implemented dispatcher. Otherwise it would get the calling assembly where the dispatcher is implemented (I suspect because the implemented dispatcher would call the IDispatcher)

- I didn't use IVoidDispatcher because I need all the dispatchers to have the same signature for Dispatch(), GetTypes() and HandleTypes() to use the LicensedDispatcher wrapper class. On the other hand, the construction of these things is getting kinda complicated. Maybe we should look into building a DispatcherFactory class?

-I made some tests demonstrating the use of [LicenseRequired].

Co-Authored-By: Frederik Madsen <FredTheDane@users.noreply.github.com>

SoftAdviceJR commented:
I tried merging our ideas.

- I moved when the callingassembly is added to the assemblies because it was causing troubles. It's now added in the constructor of the implemented dispatcher. Otherwise it would get the calling assembly where the dispatcher is implemented (I suspect because the implemented dispatcher would call the IDispatcher)

- I didn't use IVoidDispatcher because I need all the dispatchers to have the same signature for Dispatch(), GetTypes() and HandleTypes() to use the LicensedDispatcher wrapper class. On the other hand, the construction of these things is getting kinda complicated. Maybe we should look into building a DispatcherFactory class?

- I made some tests demonstrating the use of [LicenseRequired].

SoftAdviceJR committed:
removed two superfluous folders

SoftAdviceJR committed:
Update ValidatorDispatcher.cs

forgot to load the CallingAssembly()

SoftAdviceJR committed:
variable name change

handlerType => configType

SoftAdviceJR commented:
I just stumbled upon this:
https://developer.m-files.com/Frameworks/Vault-Application-Framework/Reference/html/75a03b3d-c244-f807-1d48-81366332dc56.htm

Which is not documented anywhere else and my colleague had also never heard of it. 

I'm going to try asking Craig how this should be used, because I can't figure it out from the reference.

But depending on how useful it is I'll have to rework the modules part of the LicensedDispatcher.

SoftAdviceJR commented:
So I just realised why constructing the dispatcher with the command is stupid. So instead I'm going to try to rework it so it can handle a list of commands.

```
[EventHandler(MFEventHandlerType.MFEventHandlerBeforeCheckInChanges)]
[EventHandler(MFEventHandlerType.MFEventHandlerBeforeCreateNewObjectFinalize)]
public void MyEventHandler(EventHandlerEnvironment env)
{
    //Make multiple dispatchers??? That seems wrong...
}
```

SoftAdviceJR committed:
dispatcher rework

I changed the dispatcher signature. Instead of passing the command to the dispatcher, each dispatcher now has a method AddCommand were any number of commands can be passed through.

SoftAdviceJR commented:
I reworked the dispatcher interface a bit. Now every dispatcher can take any number of commands. This is useful for event handlers for multiple events.

FredTheDane commented:
> 
> 
> I reworked the dispatcher interface a bit. Now every dispatcher can take any number of commands. This is useful for event handlers for multiple events.

A lot of new stuff going on here. Can see that you have struggled a little. My apologies for changing the dispatchers to static and then back again the day after.
Should get a chance to look at this today or tomorrow :)

FredTheDane commented:
@SoftAdviceJR my main concern with the internal state of "Commands" in dispatchers is that this will require a new instantiation of Dispatchers each time an event handler is called, rather than, for example, a single instantiation of a Dispatcher in a VaultApplication that can then be called from all event handlers, while still maintaining a TypeCache for all command types such that the concrete types will not have to be found on every command execution.

For multiple commands execution, rather than using the "AddCommand" method, instead try and implement the Dispatch method to have a signature such as this:

`public abstract TReturn Dispatch(params ICtrlVAFCommand[] commands);`

This way, you may still parse multiple commands to a single dispatcher, while not having to have an internal Command List state. Also do try and utilize the TypeCache as much as possible for Command -> Concrete Type mapping, as it will save a lot of time searching for things that do not change :)

Otherwise fantastic work on the licensing attributes! 🚀

SoftAdviceJR commented:
You're right of course. I was trying to avoid passing commands to the dispatch method, because it would required them from the validatordispatcher and the backgrounddispatcher. But I forgot that using params would give you the option of just not passing any commands. I also moved the exceptionhandler to the constructor, but this might also be better fit for the dispatch method. 
I was just tunnel visioning on keeping the dispatch method clean of all arguments.

I'll implement the changes and rewrite the tests.

SoftAdviceJR commented:
Adding the exceptionhandler to the method signature seems quite bothersome, because you have to start specifying parameter names e.g.

`dispatcher.Dispatch(commands: command_1)`

Maybe it's better to set that as a class property?

FredTheDane commented:
> 
> 
> Adding the exceptionhandler to the method signature seems quite bothersome, because you have to start specifying parameter names e.g.
> 
> `dispatcher.Dispatch(commands: command_1)`
> 
> Maybe it's better to set that as a class property?

Could be, yeah.

I did not really have any specific reasoning behind having it in the Dispatch method other than being able to treat exceptions differently in one event handler / location in code compared to another. 

Having it as a class property and not a part of the dispatch method signature should be just fine 👍

SoftAdviceJR commented:
nvm, solved it like this:
```c#
public TReturn Dispatch(bool throwExceptions = false, Action<Exception> exceptionHandler = null, params ICtrlVAFCommand[] commands)
        {
            try
            {
                return Dispatch(commands);
            }
            catch(Exception ex)
            {
                if(exceptionHandler != null)
                {
                    exceptionHandler(ex);
                }
                if (throwExceptions)
                {
                    throw ex;
                }
            }

            return default;
        }

public abstract TReturn Dispatch(params ICtrlVAFCommand[] commands);
```

SoftAdviceJR committed:
Changed Dispatch() signature

dispatcher.Dispatch() => dispatcher.Dispatch(params ICtrlVAFCommands[] commands)

Also the CommandDispatcher has been renamed to EventDispatcher for clarity, but has regained the void Dispatch() signature.

SoftAdviceJR committed:
Updated the ValidatorDispatcher

Any dispatcher with a return type now also has a results cache
Added a speedtest for the ValidatorDispatcher to demonstrate.

Also added the three dispatchers as properties to the ConfigurableVaultApplication, so users don't have to worry about instantiation.
To make them LicensedDispatchers, users just add the [UseLicensing] attribute on the VaultApplication class

SoftAdviceJR committed:
moved ConfigurableVaultApplicationBase

SoftAdviceJR committed:
Fixed an issue where on a live server, no assemblies were found for the validator/background dispatchers

SoftAdviceJR commented:
I think the only thing I'm really missing right now is having a handler than can cope with multiple commands. Something like:

```c#
class MultipleCommandHandler: 
    CommandHandler<BeforeCheckInChangesCommand<Configuration>>, 
    CommandHandler<BeforeCreateNewObjectFinalizeCommand<Configuration>>
{
    //
}
```

I'll look into that tomorrow.

SoftAdviceJR committed:
changed ICommandHandler to an interface

Now it can be inherited multiple times. You just implement Handle() multiple times.
You can use on Handle to cast the command and then call another one.

SoftAdviceJR committed:
Update EventDispatcher.cs

SoftAdviceJR commented:
Now a handler can handle multiple commands. I think that's everything I'm going to need for a while.

I implemented it in a way where if multiple commands call the same handler in one dispatch call, it only handles the first command. It seemed that the same handler being called multiple times might've been unexpected or unwanted, but I'm not sure.

I've written two tests to demonstrate.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Validators/ValidatorDispatcher.cs:
```diff
@@ -0,0 +1,144 @@
+
+        }
+
+        //private object GetConfigPropertyOfType(object config, Type configSubType)
```
Is this specific code still needed?

SoftAdviceJR replied:
Not really, I just didn't have the heart to get rid of it yet. But there's a similar method in BackgroundDispatcher (which should probably also be removed to just work with the top level Configuration) so I'll remove it.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Validators/ValidatorDispatcher.cs:
```diff
@@ -0,0 +1,144 @@
+
+            foreach (Type concreteValidator in concreteValidators)
+            {
+                if (ResultsCache.TryGetValue(concreteValidator, out IEnumerable<ValidationFinding> findings))
```
I would suggest the following modification to this loop:

```
                 if (!ResultsCache.TryGetValue(concreteValidator, out IEnumerable<ValidationFinding> findings))
                {
                    var concreteHandler = Activator.CreateInstance(concreteValidator) as ICustomValidator;

                    findings = concreteHandler.Validate(validatorCommand);

                    ResultsCache.TryAdd(concreteValidator, findings);
                }

                return findings;
```

Unless of course there is a particular reason to looping over the findings and yield returning them individually in the end :)

FredTheDane replied:
Hmm, perhaps there is a reason to yield returning in the end. I am a little unsure myself.
But at least, do consider inverting the if statement as in the suggestion above

SoftAdviceJR replied:
I inverted the if statement.

I tested replacing the yield statements with return statements but it slowed down the speed tests, so I kept it for now.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/Dispatcher.cs:
```diff
@@ -0,0 +1,88 @@
+            }
+            catch (Exception ex)
+            {
+                if (exceptionHandler != null)
```
I agree, this approach is probably better than having a boolean to determine whether or not to throw on exceptions in addition to also handling exceptions.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/Dispatcher_Common.cs:
```diff
@@ -0,0 +1,59 @@
+        {
+            Assemblies.AddRange(assemblies);
+            //Never search CtrlVAF.Core
+            Assemblies.RemoveAll(a => a == GetType().Assembly);
```
Ohwait, so this removes any assemblies from the list that are a part of CtrlVAF.Core?

SoftAdviceJR replied:
Yeah. When adding CallingAssembly() it would often add CtrlVAF.Core. Since that should never be a use case I thought this might be a good improvement.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/LicensedDispatcher.cs:
```diff
@@ -0,0 +1,171 @@
+        public LicensedDispatcher(Dispatcher dispatcher, LicenseContentBase license = null)
+        {
+            this.license = license;
+            this.dispatcher = dispatcher;
```
Any particular reason for decorating the underlying dispatcher?
As far as I can see, the dispatcher.GetTypes should return the same as GetTypes of the licenseddispatcher.
Likewise, the handleConcreteTypes could simply not be overriden.

The LicensedDispatcher is, in itself, already a full Dispatcher.

SoftAdviceJR replied:
LicensedDispatcher.GetTypes() calls the wrapped dispatcher's GetTypes (be it background validator or event) to get the types it should resolve and then filters those types based on the license and [LicenseRequired] attribiutes. LicensedDispatcher inherits from the abstract Dispatcher so it has to override the HandleConcreteTypes() method.

It seemed like an extensible way to implement it.

FredTheDane commented on CtrlVAF/CtrlVAF.AdditionalAssembly/TestHandler.cs:
```diff
@@ -1,22 +1,24 @@
-        }
-    }
-}
+﻿using CtrlVAF.Commands.Commands;
```
Is this testhandler actually used anywhere?
If not, lets just get rid of the AdditionalAssembly project for now :))

SoftAdviceJR replied:
It's used in one of your tests. It was in CtrlVAF.Core at first, I just moved it to a different assembly altogether to keep CtrVAF.Core cleaner.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/ConfigurableVaultApplicationBase.cs:
```diff
@@ -0,0 +1,93 @@
+        {
+            var command = new ValidatorCommand<TSecureConfiguration> { Vault = vault, Configuration = config };
+
+            return ValidatorDispatcher.Dispatch(command);
```
Should the base validation method be called as well, or is that not necessary?

SoftAdviceJR replied:
I don't think so. It's also weird to add because I already return the dispatch results.

FredTheDane replied:
I was just thinking if the base method also did some validation that you would want to be returned. For instance, in the example on the developer portal, the base method is called [https://developer.m-files.com/Frameworks/Vault-Application-Framework/Configuration/Custom-Validation/](url)

The things returned by the base call could just be concattenated to the once retrieved by the validatordispatcher and then returned in the end.

But again, I am unsure if the base call returns any findings of value.

SoftAdviceJR replied:
I don't think it does. But I'll just concatenate them then.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/LicensedDispatcher.cs:
```diff
@@ -0,0 +1,171 @@
+        public LicensedDispatcher(Dispatcher<TReturn> dispatcher, LicenseContentBase license = null)
+        {
+            this.license = license;
+            this.dispatcher = dispatcher;
```
Same comment as below :))

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/LicensedDispatcher.cs:
```diff
@@ -0,0 +1,171 @@
+            return HandleConcreteTypes(types, commands);
+        }
+
+        protected internal override IEnumerable<Type> GetTypes(params ICtrlVAFCommand[] commands)
```
This method to me looks like exactly the same as the other GetTypes method in the alternative LicensedDispatcher class below.
I would probably consider including this method in a LicensedDispatcher_Common abstract class and have both LicensedDispatcher inherit from this.

SoftAdviceJR replied:
I can't implement them commonly because their return types are different. At least that's what I think. 
This way I thought it was cleaner to have them seperately instead of void Dispatch() methods needing an object return type.

FredTheDane replied:
The return type of the GetTypes method is the same both here and on line 118. The code is exactly the same

SoftAdviceJR replied:
nvm I thought you meant the HandleConcreteTypes method for some reason. The gettypes is exactly the same I think yeah. I'll extract that then.

SoftAdviceJR replied:
Ah the issue is that LicensedDispatcher already inherits from Dispatcher and you can't inherit from two classes at once. 

I created a static helper class to implement the functionality of the filter.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/EventHandlers/EventDispatcher.cs:
```diff
@@ -0,0 +1,136 @@
+                dispatchableHandlerTypes.AddRange(handlerTypes);
+            }
+
+            return dispatchableHandlerTypes;
```
I suppose that here there is a slim chance of duplicate handlerTypes being returned?
As multiple commands of the same type could be dispatched in the beginning.
Perhaps, a distinct call should be placed here?

SoftAdviceJR replied:
I prefer keeping a log of handled command types then. It seems more efficient than adding duplicates and then removing them.

FredTheDane commented on CtrlVAF/CtrlVAF.Core/EventHandlers/EventDispatcher.cs:
```diff
@@ -0,0 +1,136 @@
+                //If the concrete types have already been retrieved and cached before, simply handle those
+                if (TypeCache.TryGetValue(commandType, out var cachedTypes))
+                {
+                    dispatchableHandlerTypes.AddRange(cachedTypes.Distinct());
```
I think the Distinct call should be moved to when handlerTypes are moved to the TypeCache on line 85.
Otherwise Distinct is called way more than necessary :)

SoftAdviceJR replied:
Oh yeah, I forgot to mention this. I was having some weird issue where the result from trygetvalue was getting me duplicates (even though the dictionary didn't contain them). But I might've gone a bit wild with the Distinct() calls. I'll look into optimizing.

FredTheDane reviewed (commented):
Please see my comments in the review :)

All in all good work!

FredTheDane commented:
> 
> 
> Now a handler can handle multiple commands. I think that's everything I'm going to need for a while.
> 
> I implemented it in a way where if multiple commands call the same handler in one dispatch call, it only handles the first command. It seemed that the same handler being called multiple times might've been unexpected or unwanted, but I'm not sure.
> 
> I've written two tests to demonstrate.

Instead of relying on implementing interfaces for eventhandling, we could use a similar approach to your licensing and simply use attributes to determine which eventhandlertypes a given class should be called on and have all commandhandler inherit a common interface.

i.e, not do this
```csharp
public class AfterSetPropertiesHandler2 : IEventHandler<AfterSetPropertiesCommand<Configuration>>
    {
        public void Handle(AfterSetPropertiesCommand<Configuration> command)
        {
            command.Env.Input = command.Configuration.Name;
        }
    }
```

and instead do something like this
```csharp
[EventTypes([MFEventHandlerType.MFEventHandlerAfterSetProperties, MFEventHandlerType.MFEventHandlerAfterCheckinChanges])]
public class AfterSetPropertiesHandler2 : IEventHandler<Configuration>
    {
        public void Handle(Command<Configuration> command)
        {
            command.Env.Input = command.Configuration.Name;
        }
    }
```

SoftAdviceJR committed:
various improvements

SoftAdviceJR commented:
I think that could be done. We would need to create a mapping and I would also suggest extending the things that are available as command properties. Though I don't know if it's something we want to do.

Right now it's possible to implement your own custom commands and just add whatever property you want, but that would become impossible. I don't know if we want to lose that.

We could allow passing custom data as a json or something, but that will introduce annoying problems again for the user (e.g. ObjVer is hard/impossible to deserialize iirc)

FredTheDane reviewed (commented):
I am happy with the changes 👍 

Once merged in I will have a look at providing documentation for the remaining modules and touching some things up here and there. After that I will provide it as a package on nuget.

Are you happy with this being merged in @SoftAdviceJR, or do you have any last changes?
Will await your response. No rush :)

SoftAdviceJR committed:
Implemented a ClearCache method

This seemed important for the validatordispatcher, who needs to clear resultscache when the configuration is updated.

SoftAdviceJR commented:
I'm getting a problem where a vault reference is deemed invalid in the validation process. So I'll need to debug that. 
After that I think I want to rewrite the background dispatcher a bit so it matches the other two more in usage.

But that's about it for now.

SoftAdviceJR committed:
Fixed ResultsCache issue

The lazy loading of the resultscache resulted in trying to use an invalid ResultsCache. This is fixed by forcing the ResultsCacheto have the actual results in cache by calling .ToList().

SoftAdviceJR committed:
Added a method to query the ResultsCache

SoftAdviceJR committed:
Reimplemented lazy loading but added call ToArray() to load all at once

There's no actual speed difference, just thought we'd want to keep the lazy loading as an easy option if we ever find a solution.

SoftAdviceJR commented:
For some reason M-Files runs the CustomValidation twice when the config is updated? No clue why though 🤷 

Also the lazy loading was interacting weirdly with the idea of a ResultsCache so I removed the lazy loading. It does add considerable time to the speedtests, so if you've got an idea on how to solve it while keeping the lazy loading, I'm all ears.

Also, maybe the idea about passing members of configuration instead of the whole thing might be worth revisiting? It could solve the problem that Damon is having with circular references? Since you'd only need a reference to the 'subType' of Configuration and not the whole thing.

SoftAdviceJR committed:
Improved TargetInvocationException handling

The stack trace was not being preserved and this solves that.

SoftAdviceJR committed:
Improved the way Configuration is set on a background task

I'd like to implement something similar for most handlers and other properties

SoftAdviceJR commented:
I'm alright with the way it's implemented now, but I have some more ideas I want to try out, so I'll start a new branch for those.

SoftAdviceJR committed:
some name changes and changes to BackgroundDispatcher

SoftAdviceJR committed:
renamed commands namespace to events

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/Dispatcher.cs:
```diff
@@ -84,5 +90,31 @@ public TReturn Dispatch(Action<Exception> exceptionHandler, params ICtrlVAFComma
+        /// <returns></returns>
+        public TReturn GetCachedResults(Type keyType)
+        {
+            return ResultsCache[keyType];
```
To ensure that what you want to get is actually there, otherwise default
```suggestion
            if(!ResultsCache.TryGetValue(keyType, out var value)) {
                return default;
            }
            return value;
```

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Core/Dispatcher.cs:
```diff
@@ -84,5 +90,31 @@ public TReturn Dispatch(Action<Exception> exceptionHandler, params ICtrlVAFComma
+        /// <returns></returns>
+        public TReturn GetCachedResults(object keyObject)
+        {
+            return ResultsCache[keyObject.GetType()];
```
Same comment as above
```suggestion
            return GetCachedResults(keyObject.GetType());
```

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Validators/ValidatorDispatcher.cs:
```diff
@@ -15,8 +15,6 @@ namespace CtrlVAF.Validators
     {
-        
-
         public ValidatorDispatcher()
```
Maybe we should just get rid of this constructor?

SoftAdviceJR replied:
In this branch it's fine to get rid of it. I have some plans for it though 😈

FredTheDane commented on CtrlVAF/CtrlVAF.Core/Validators/ValidatorDispatcher.cs:
```diff
@@ -27,7 +25,7 @@ public override IEnumerable<ValidationFinding> Dispatch(params ICtrlVAFCommand[]
             var types = GetTypes(commands);
 
-            return HandleConcreteTypes(types, commands);
+            return HandleConcreteTypes(types, commands).ToArray();
```
Hmm, this must be a consequence of the way findings are returned from HandleConcreteTypes. Without it, the code will likely continue to execute and perhaps finish before all findings have been yield returned back.

I think this is ok for now

FredTheDane reviewed (commented):
See comments below 👍 
Was able to run all tests just fine too.

Really looking forward to using this in my own projects

SoftAdviceJR committed:
Added sanity check to getting cached results

Co-authored-by: Frederik Madsen <FredTheDane@users.noreply.github.com>

SoftAdviceJR committed:
DRY on get cached results

Co-authored-by: Frederik Madsen <FredTheDane@users.noreply.github.com>

SoftAdviceJR committed:
removed constructor for ValidatorDispatcher

SoftAdviceJR committed:
Merge branch 'Licensing' of https://github.com/SoftAdviceJR/CtrlVAF into Licensing

FredTheDane commented:
Good work @SoftAdviceJR !
